### PR TITLE
fix(client): exports map

### LINKED ISSUE #1984: Prisma instrumentation no longer works on >5.9.0
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
Prisma just released [5.9.0](https://github.com/prisma/prisma/releases/tag/5.9.0).  It looks like they restructured exports to where instrumenting `@prisma/client` is not sufficient.  

## Expected Behavior

Instrumentation of prisma 5.9.0+ will still function

## Steps to Reproduce

```sh
cd test/versioned/prisma
npm i prisma@latest @prisma/client@5.9.0 --no-save --no-package-lock
node prisma.tap.js
```

## Additional context
This will be an exercise in figuring out the appropriate paths to give to `shimmer.registerInstrumentations`. I've tried the following so far and it's very hard to determine the _right_ paths.  [WIP branch](https://github.com/bizob2828/node-newrelic/commit/b37e4a75416db9038d341654eb4b4e017acacbf1)

You can see the export changed

5.8.1 with debug on in require-in-the-middle

```sh
2024-01-30T16:54:52.135Z require-in-the-middle resolved filename to module: @prisma/client (id: @prisma/client, resolved: @prisma/client, basedir: /Users/revans/code/release/node-newrelic/test/versioned/prisma/node_modules/@prisma/client)
2024-01-30T16:54:52.136Z require-in-the-middle calling require hook: @prisma/client
```

5.9.0

```sh
2024-01-30T16:39:22.506Z require-in-the-middle resolved filename to module: @prisma/client (id: @prisma/client, resolved: @prisma/client/default, basedir: /Users/revans/code/pull-requests/node-newrelic/test/versioned/prisma/node_modules/@prisma/client)
2024-01-30T16:39:22.506Z require-in-the-middle ignoring require of non-main module file: /Users/revans/code/pull-requests/node-newrelic/test/versioned/prisma/node_modules/@prisma/client/index.js
```

### LINKED ISSUE #22875: Error after Upgrading from 5.8.1 to 5.9.0
### Bug description

After upgrading the following error is shown:
```
Error: Cannot find module 'F:\XXX\XXX\backend\node_modules\@prisma\client\runtime\library'
    at createEsmNotFoundErr (node:internal/modules/cjs/loader:1060:15)
    at finalizeEsmResolution (node:internal/modules/cjs/loader:1053:15)
    at resolveExports (node:internal/modules/cjs/loader:538:14)
    at Function.Module._findPath (node:internal/modules/cjs/loader:607:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1025:27)
    at Function.Module._load (node:internal/modules/cjs/loader:885:27)
    at Module.require (node:internal/modules/cjs/loader:1105:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (F:\dev\ecence\backend\src\models\user\user.service.ts:7:1)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
```

### Expected behavior

Run Application. Works fine on 5.8.1.

I guess this error occurs because of the following line in our nestjs application:
```
import { PrismaClientKnownRequestError } from '@prisma/client/runtime/library';
```

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: Windows
- Database: MySQL 8
- Node.js version: 18.13.0

### Prisma Version

```
v5.9.0
```

### LINKED ISSUE #22877: [5.9.0] `@prisma/client` in Next.js middleware
### Bug description

Using `@prisma/client` at runtime in a Next.js middleware is not working anymore after updating to 5.9.0.

### How to reproduce

Follow the instructions in this minimal reproduction:
https://github.com/darthmaim-reproductions/prisma-prisma-22877

### Expected behavior

I should be able to use `@prisma/client` in a Next.js middleware.

This was working in 5.8.1, which can be verified by changing the prisma version in the `package.json` of the reproduction and following the reproduction steps. Now `User` ist logged in the terminal, instead of the error occuring in 5.9.0

### Prisma information

See repository for prisma setup

### Environment & setup

- OS: Ubuntu inside WSL 2
- Database: doesn't matter
- Node.js version v20.7.0

### Prisma Version

```
prisma                  : 5.9.0
@prisma/client          : 5.9.0
Computed binaryTarget   : debian-openssl-1.1.x
Operating System        : linux
Architecture            : x64
Node.js                 : v20.7.0
Query Engine (Node-API) : libquery-engine 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Schema Engine           : schema-engine-cli 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules/@prisma/engines/schema-engine-debian-openssl-1.1.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.9.0-32.23fdc5965b1e05fc54e5f26ed3de66776b93de64
Default Engines Hash    : 23fdc5965b1e05fc54e5f26ed3de66776b93de64
Studio                  : 0.497.0
```

### LINKED ISSUE #22884: [v5.9.0] `$extends` always return `any`
### Bug description

Basically the title. `$extends` will always return `any` instead of the generated `PrismaClient` type. This issue is not reproducible on v5.8.1.

### How to reproduce

[Edit Prisma extends always return any type](https://codesandbox.io/p/devbox/prisma-extends-always-return-any-type-dwx89z?file=%2Fpackage.json&embed=1)

### Expected behavior

`$extends` should return a valid `PrismaClient` type.

### Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Test {
  id   String @id
}
```

```ts
import { PrismaClient } from "@prisma/client";
import { readReplicas } from "@prisma/extension-read-replicas";

// this has the generated type
const client = new PrismaClient();

// this will always in type `any`
const clientWithExtends = new PrismaClient().$extends();

// same applies with a valid extension specified
const clientWithExtends = new PrismaClient().$extends(
  readReplicas({ url: "" }),
);
```

### Environment & setup

- OS: Code Sandbox
- Database: PostgreSQL
- Node.js version: 20.9.0

### Prisma Version

```
prisma                  : 5.9.0
@prisma/client          : 5.9.0
Computed binaryTarget   : debian-openssl-3.0.x
Operating System        : linux
Architecture            : x64
Node.js                 : v20.9.0
Query Engine (Node-API) : libquery-engine 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules/.pnpm/@prisma+engines@5.9.0/node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules/.pnpm/@prisma+engines@5.9.0/node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.9.0-32.23fdc5965b1e05fc54e5f26ed3de66776b93de64
Default Engines Hash    : 23fdc5965b1e05fc54e5f26ed3de66776b93de64
Studio                  : 0.497.0
```

### LINKED ISSUE #22885: please how i cant fix this message issues
```
Server Error
Error: Prisma Client is unable to run in an edge runtime. As an alternative, try Accelerate: https://pris.ly/d/accelerate.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues/new

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Source
lib/db.ts (7:39) @ eval

   5 | }
   6 |
>  7 | export const db = globalThis.prisma || new PrismaClient();
     |                                       ^
   8 |
   9 | if (process.env.NODE_ENV !== 'production') globalThis.prisma = db;
  10 |
  ```

### LINKED ISSUE #22886: Prisma edge runtime error
`Error: Prisma Client is unable to run in an edge runtime. As an alternative, try Accelerate`

 when I write auth.config,ts file so it's shown me this error

### LINKED ISSUE #22889: `Error: Prisma Client is unable to run in an edge runtime. As an alternative, try Accelerate: https://pris.ly/d/accelerate.`
Screenshot 2024-01-31 123922

### LINKED ISSUE #22893: [5.9.0] Error: Prisma Client is unable to run in an edge runtime. As an alternative, try Accelerate: https://pris.ly/d/accelerate.
I'm using Next Auth with next.js 14

Screenshot 2024-01-31 213240

### PR DESCRIPTION
**Pull request by millsp**
Fixes https://github.com/newrelic/node-newrelic/issues/1984
Closes https://github.com/prisma/prisma/issues/22875
Closes https://github.com/prisma/prisma/issues/22877
Closes https://github.com/prisma/prisma/issues/22889
Closes https://github.com/prisma/prisma/issues/22886
Closes https://github.com/prisma/prisma/issues/22885
Closes https://github.com/prisma/prisma/issues/22884
Closes https://github.com/prisma/prisma/issues/22893

/integration

### CODE REVIEW

File: packages/client/package.json [hunk 1]
```diff
@@ -83,7 +83,14 @@
...
     },
-    "./*": "./*"
+    "./*.js": "./*.js",
+    "./*.d.ts": "./*.d.ts",
+    "./*": {
+      "types": "./*.d.ts",
+      "require": "./*.js",
+      "import": "./*.js",
+      "default": "./*.js"
+    }
```
Comment (millsp) [inline packages/client/package.json:93]: This is to fix `https://github.com/prisma/prisma/issues/22875`. Also thanks to that, we can remove the `.js` extensions we just added on the types.

File: packages/client/src/__tests__/types/types.test.ts [hunk 2]
```diff
@@ -58,6 +58,7 @@ describe('valid types', () => {
 async function runTsd(dir: string) {
   const diagnostics = await tsd({
     cwd: dir,
+    typingsFile: 'index.d.ts',
```
Comment (millsp) [inline packages/client/src/__tests__/types/types.test.ts:61]: `tsd` wants to import `default.d.ts` but our test is via `index.d.ts` in that folder.

File: packages/client/package.json [hunk 3]
```diff
@@ -29,9 +29,9 @@
     "SQLServer",
     "MongoDB"
   ],
-  "main": "index.js",
+  "main": "default.js",
   "browser": "index-browser.js",
-  "types": "index.d.ts",
+  "types": "default.d.ts",
```
Comment (millsp) [inline packages/client/package.json:34]: This is an attempt to fix https://github.com/newrelic/node-newrelic/issues/1984. I don't think there was anything particularly wrong with `index` here, but rather a bug or limitation with `require-in-the-middle` (used by New Relic). 

There's an interplay between `main` being different from `exports` and that makes the `Hook` of `require-in-the-middle` not work when that is the case, although `require` did load the module correctly via the exports map and not `main`. 

In other words, `Hook` is supposed to trigger when `require('@prisma/client')` is executed, but it does not if `main` and `exports` point to different files, although in the end it does not matter. Maybe an edge case or a bug.

File: packages/client/package.json (hunk 1 above)
Comment (millsp) [inline packages/client/package.json:93]: This is to fix `https://github.com/prisma/prisma/issues/22875`. The exports map made imports such as `@prisma/client/runtime/library` impossible and needed a `.js` extension to work. Now it allows both with and without `.js`.

An exports map reads from top to bottom, with most specific keys first taking precedence over less specific ones below. So here we are basically saying: "allow imports in `.js`, `.d.ts`, and in any other case where it is not specified (eg. `/runtime/library`), then add the `.js` extension automatically.

It will allow people to continue using internals for the time being, we can decide if we want to prevent this in Prisma 6.

File: packages/client/src/generation/generateClient.ts [hunk 4]
```diff
@@ -129,8 +129,8 @@ export async function buildClient({
 
   const pkgJson = {
     name: getUniquePackageName(datamodel),
-    main: 'index.js',
-    types: 'index.d.ts',
+    main: 'default.js',
+    types: 'default.d.ts',
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:133]: Same as https://github.com/prisma/prisma/pull/22880/files#r1472081850

File: packages/client/package.json (hunk 1 above)
Comment (millsp) [reply packages/client/package.json:93]: Updates https://github.com/newrelic/node-newrelic/issues/1984#issuecomment-1918097671

File: packages/client/package.json (hunk 3 above)
Comment (bizob2828) [reply packages/client/package.json:34]: also `require-in-the-middle` is used by opentelemetry js, datadog, elastic to name a few, it will affect them as well

Comment (millsp) [reply packages/client/package.json:34]: Hey, could you please give `5.10.0-integration-fix-client-exports-map.8` a try?

Comment (millsp) [reply packages/client/package.json:34]: (Nothing should have fundamentally changed)

Comment (bizob2828) [reply packages/client/package.json:34]: `5.10.0-integration-fix-client-exports-map.8` still works. I think as long as the main matches the exported default it's a sufficient enough fix.

Comment (millsp) [reply packages/client/package.json:34]: Thanks!

File: packages/client/src/generation/generateClient.ts [hunk 5]
```diff
@@ -178,8 +175,8 @@ export async function buildClient({
     fileMap['wasm.js'] = await JS(wasmClient)
     fileMap['wasm.d.ts'] = await TS(wasmClient)
   } else {
-    fileMap['wasm.js'] = await fs.readFile(path.join(scriptsDir, 'wasm-da-feature-deactivated.js'), 'utf-8')
-    fileMap['wasm.d.ts'] = await fs.readFile(path.join(scriptsDir, 'wasm-da-feature-deactivated.d.ts'), 'utf-8')
+    fileMap['wasm.js'] = fileMap['index-browser.js']
+    fileMap['wasm.d.ts'] = fileMap['default.d.ts']
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:179]: Restores the browser version to be used in edge functions, as it was before.

File: packages/client/tests/e2e/deprecated-prisma-client/src/default.ts [hunk 6]
```diff
@@ -1,3 +1,9 @@
 import { PrismaClient } from '@prisma/client'
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import { PrismaClientKnownRequestError as E1 } from '@prisma/client/runtime/library'
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import { PrismaClientKnownRequestError as E2 } from '@prisma/client/runtime/library.js'
```
Comment (millsp) [inline packages/client/tests/e2e/deprecated-prisma-client/src/default.ts:5]: Tests that these are valid paths.

File: packages/client/package.json [hunk 7]
```diff
@@ -83,6 +83,26 @@
...
+    "./runtime/edge": {
+      "require": "./runtime/edge.js",
+      "import": "./runtime/edge.js",
+      "default": "./runtime/edge.js"
+    },
+    "./runtime/wasm": {
+      "require": "./runtime/wasm.js",
+      "import": "./runtime/wasm.js",
+      "default": "./runtime/wasm.js"
+    },
```
Comment (millsp) [inline packages/client/package.json:105]: I initially had a more compact version, which turned out to be a time sink to make it work with Next.js 13 (misinterpreted), so I opted towards a more verbose and more archaic version of it.

```ts
    "./*.js": "./*.js",
    "./*.d.ts": "./*.d.ts",
    "./*": {
      "types": "./*.d.ts",
      "require": "./*.js",
      "import": "./*.js",
      "default": "./*.js"
    }
```

File: packages/client/package.json [hunk 8]
```diff
@@ -83,6 +83,30 @@
...
+      "require": "./runtime/edge.js",
+      "import": "./runtime/edge.js",
+      "default": "./runtime/edge.js"
+    },
+    "./runtime/wasm": {
+      "types": "./runtime/wasm.d.ts",
+      "require": "./runtime/wasm.js",
+      "import": "./runtime/wasm.js",
+      "default": "./runtime/wasm.js"
+    },
```
Comment (millsp) [inline packages/client/package.json:109]: Allows importing without specifying the extension.

File: packages/client/tests/e2e/prisma-client-imports/src/default.ts
Comment (millsp) [inline packages/client/tests/e2e/prisma-client-imports/src/default.ts:1]: Not really a new file, but side effect of renaming the folder.

File: packages/client/package.json (hunk 8 above)
Comment (millsp) [reply packages/client/package.json:109]: I initially had a more compact version (see below), which turned out to be a time sink to make it work with Next.js 13 (misinterpreted in 13 but works in 14), so I opted towards a more verbose and more archaic version of it.

```ts
    "./*.js": "./*.js",
    "./*.d.ts": "./*.d.ts",
    "./*": {
      "types": "./*.d.ts",
      "require": "./*.js",
      "import": "./*.js",
      "default": "./*.js"
    }
```

File: packages/client/tests/e2e/prisma-client-imports/tests/main.test.ts [hunk 9]
```diff
@@ -93,7 +93,9 @@ function isPrismaClientDeprecated(fileName: string, options: any) {
 
 function assertNoErrors(errors: readonly ts.Diagnostic[]) {
   if (errors.length > 0) {
-    throw new Error(`TS errors: ${errors.map((err) => err.messageText).join('\n')}`)
+    errors.map(console.error)
+
+    throw new Error(`Test exited with ${errors.length} errors. See above for details.`)
```
Comment (millsp) [inline packages/client/tests/e2e/prisma-client-imports/tests/main.test.ts:98]: Errors are objects so they are easier to read like this (if they happen)

File: packages/client/tests/e2e/deprecated-prisma-client/readme.md
Comment (millsp) [inline packages/client/tests/e2e/deprecated-prisma-client/readme.md:1]: Not deleted, just moved.

**Review by Jolg42 — APPROVED** [review]
Nice one, thanks for renaming the e2e tests 🙌🏼

File: packages/client/package.json (hunk 8 above)
Comment (millsp) [inline packages/client/package.json:109]:
```suggestion
```
